Add RxRedux sample

artem-zinnatullin opened the pull request:
PR adds sample of Domic integration with [RxRedux](https://github.com/freeletics/RxRedux) made by @sockeqwe at Freelitics

Please note that while I seem to understand ideas of Redux and some pros/cons, it might be not very idiomatic.

General thoughts on Domic + Redux/MVI:

- Domic's efficient rendering with automatic state diffing is playing really well with Redux, you can run the app and see yourself
- Domic becomes nice unified way of providing observables for `inputActions` produced by View layer, however not much profit here compared to RxBinding
- Domic's `test` module becomes useless unless you want to use it for functional in-memory tests for your Redux app on JVM

General thoughts on Redux:

- I've faced quite a lot of boilerplate/productivity problems, mainly this is due to not being able to `copy()` previous state to produce new one
- Sharing some logic for similar actions becomes strange, see `ChangeEmail` and `ChangePassword` action handling in state machine
- Redux generally is very verbose
- I do feel pretty safe because compiler forces me to check a lot of states because of sealed classes everywhere
- However it's still not checking all combinations of possible states and if you try to write code that way it just becomes crazy to reason about state combinations

Some thoughts on RxRedux:

- It's nice, I find it easier to reason about than [Mobius](https://github.com/spotify/mobius), I guess mainly because I'm comfortable with RxJava and RxRedux is  built directly on top of it
- I'm still not sure if `StateAccessor` was necessary, compared to just passing current state as parameter, I'll have to try in more complicated scenarios when something can happen in parallel to async action I guess
- Kotlin's typealiases are finally useful lol

cc @sockeqwe, code review from you would be reaaaaaly appreciated :)

artem-zinnatullin commented:
Added a bit more about Redux in general

ming13 commented on samples/redux/rxredux/src/test/java/com/lyft/samples/redux/rxredux/signin/SignInStateMachineTest.kt:
```diff
@@ -0,0 +1,82 @@
+    @Test
+    fun `displays success ui`() {
+        inputActions.accept(SignInAction.ChangeEmail("test@email"))
+        inputActions.accept(SignInAction.ChangePassword("passw0rd"))
```
You can at least use JUnit 5 nesting.

artem-zinnatullin replied:
Honestly, I don't have time to deal with JUnit5 setup

artem-zinnatullin commented on samples/redux/rxredux/src/main/java/com/lyft/domic/samples/redux/rxredux/signin/AndroidSignInView.kt:
```diff
@@ -0,0 +1,74 @@
+    private val signInButton = AndroidButton(root.findViewById(R.id.sign_in_button), renderer)
+    private val resultTextView = AndroidTextView(root.findViewById(R.id.sign_result_text_view), renderer)
+
+    override val actions: Observable<SignInAction> = Observable
```
Ohh, definitely, I was like there is a way, but didn't want to spend too much time on it, will refactor, thanks

artem-zinnatullin replied:
foxed, thx

artem-zinnatullin commented on samples/redux/rxredux/src/main/java/com/lyft/domic/samples/redux/rxredux/signin/SignInAction.kt:
```diff
@@ -0,0 +1,10 @@
+    data class ChangePassword(val password: CharSequence) : SignInAction()
+    object SignIn : SignInAction()
+    object ShowSigningInUi : SignInAction()
+    object ShowSignInSuccessfulUi : SignInAction()
```
Yeah, not happy about it either : ( 

Do you have other names on mind?

artem-zinnatullin replied:
hm, I thought Actions should be verbs, but ok

artem-zinnatullin commented on samples/redux/rxredux/src/main/java/com/lyft/domic/samples/redux/rxredux/signin/SignInAction.kt:
```diff
@@ -0,0 +1,10 @@
+    object SignIn : SignInAction()
+    object ShowSigningInUi : SignInAction()
+    object ShowSignInSuccessfulUi : SignInAction()
+    data class ShowSignInFailureUi(val cause: Throwable) : SignInAction()
```
Can be logged or transformed to particular text (then it should be done not in View layer though)

artem-zinnatullin replied:
fair, I'll keep it for now because it's how it's done in other samples, created issue to track it #34

jamolkhon commented:
If you simply pass the state it may change by the time you use it. StateAccessor is needed to access the current version of the state.

artem-zinnatullin commented:
Yeah I understand, that's why I said that I probably need more real life experience with long running async actions

At the same time this kinda breaks the idea of functional pureness a bit, at least for me, maybe it's pragmatic tho

sockeqwe commented on samples/redux/rxredux/src/test/java/com/lyft/samples/redux/rxredux/signin/SignInStateMachineTest.kt:
```diff
@@ -0,0 +1,82 @@
+
+        inputActions.accept(SignInAction.SignIn)
+
+        service.signIn.accept(SignInResult.Success)
```
Nitpicking since it's not the aim of this repo to show best practices on testing but this feels super strange to me  in combination with 

```kotlin
class TestSignInService : SignInService()
       ...

        override fun signIn(credentials: SignInService.Credentials): Observable<SignInService.SignInResult> {
            signInCredentialsRelay.accept(credentials)
            return signIn
        }
}
```

especially this part:

```kotlin
signInCredentialsRelay.accept(credentials)
return signIn
```

Theoretically, if you run your application with `TestSignInService` instead of `RealSignInService`, does this even work?  I'm not sure if I like it but again, it's about domic not about testing :)

artem-zinnatullin replied:
Wait, what's wrong with it? How would you simulate an emission from the rx based business logic?

sockeqwe replied:
Sorry, I just realiized that I commented on the wrong source file.
There is nothing wrong with `service.signIn.accept(SignInResult.Success)`. Just `TestSignInService` implementation feels strange (not wrong, just strange 😄 )

sockeqwe commented on samples/redux/rxredux/src/main/java/com/lyft/domic/samples/redux/rxredux/signin/SignInStateMachine.kt:
```diff
@@ -0,0 +1,92 @@
+                                is SignInService.SignInResult.Error -> SignInAction.ShowSignInFailureUi(result.cause)
+                            }
+                        }
+                        .startWith(SignInAction.ShowSigningInUi)
```
Micro Optimization: Not sure if the `startWith()` is needed because `Reducer` could  already produce the right state just by getting `SignInAction.SignIn`. See duplicated code in reducer:

```kotlin
is SignInAction.SignIn -> SignInState.SigningIn(
                email = state.email,
                password = state.password,
                signInButtonEnabled = false
        )
        is SignInAction.ShowSigningInUi -> SignInState.SigningIn(
                email = state.email,
                password = state.password,
                signInButtonEnabled = false
        )
```

artem-zinnatullin replied:
Good catch, I wrote this side-effect before the reducer and didn't notice that logic was duplicated after that

Interestingly, both `distinctUntilChanged()` on the `reduxStore` and Domic's diffing make stuff like this work without screen flickering and go unnoticed

sockeqwe commented:
Great job 👍 

Some thoughts on your thought 😸 

> - Domic's test module becomes useless unless you want to use it for functional in-memory tests for your Redux app on JVM

As I (usually) prefer integration test over unit tests, I still see some value in having functional in-memory tests. But yeah, you could just trigger actions directly, it just a matter of where (at which layer) your functional in-memory tests start. Same for "rendering state". Of course you could just test your state machine but if its easy and cheap to test rendering on UI too (which seems to be the case with domic's test), then why not include this layer too?  I think its `MainActivityTest` vs. `SignInStateMachineTest` where `MainActivityTest` also includes actually testing real `SignInStateMachine`.


> - I've faced quite a lot of boilerplate/productivity problems, mainly this is due to not being able to copy() previous state to produce new one

Not sure what you mean? I mean, you do that in reducer in you code?!?


> - Sharing some logic for similar actions becomes strange, see ChangeEmail and ChangePassword action handling in state machine
> - Redux generally is very verbose

I feel your pain. I think this sample (it's essentially just doing a http request) is simple enough that it actually does not necessary need previous state to compute next state and therefore no state machine or redux store (or .scan() operator) at all. Hence it might feel a bit more boilerplate or verbose  or even overkill compared to your mvvm implementation. Imho its because RxJava actually is already somehow a state machine with `onNext` and `onError` and thats all you need in this example.


> - I'm still not sure if `StateAccessor` was necessary, compared to just passing current state as parameter, I'll have to try in more complicated scenarios when something can happen in parallel to async action I guess

Right, we weren't happy with this either and it caused me holding back open sourcing RxRedux for quite some time. We were looking for a better solution to that problem but eventually we didn't find a better way (if you know a better way I would be very happy to hear it). Eventually we decided to open source it with `StateAccessor` ... 
Passing just current state in as a parameter might be enough in most cases but yeah in more complex / async scenarios you might need a way to grab the latest state at any given point in time which you can do with `StateAccessor`. 

> - Kotlin's typealiases are finally useful lol

🙏

artem-zinnatullin commented on samples/redux/rxredux/src/main/java/com/lyft/domic/samples/redux/rxredux/signin/SignInAction.kt:
```diff
@@ -0,0 +1,10 @@
+package com.lyft.domic.samples.redux.rxredux.signin
+
+sealed class SignInAction {
```
Oh that's a nice idea

artem-zinnatullin replied:
Could get the actions private atm, but moving them to the StateMachine allowed to have better class name, so they're now just `State` and `Action` which looks much cleaner with all other advices you and Artur gave

artem-zinnatullin replied:
*Could not get

artem-zinnatullin commented on samples/redux/rxredux/src/main/java/com/lyft/domic/samples/redux/rxredux/signin/SignInState.kt:
```diff
@@ -0,0 +1,33 @@
+package com.lyft.domic.samples.redux.rxredux.signin
+
+sealed class SignInState(
+        open val email: CharSequence,
```
fair

artem-zinnatullin commented:
I've addressed your feedback and actually like it much more now, especially due to `abstract` instead of `open` properties in `State` class and some renamings, boilerplate reduced quite a lot. It still bothers me in the `reducer()`, but I might get used to it.

Overall I like the integration, it definitely doesn't feel anywhere near as native as MVVM + Domic, but it's not bad at all. Especially after addressing your feedback.

I must try the DSL-based approach as well though, to better understand the feel of React + Redux. A lot of people are (and were) super hyped about that exact combination for *some* reason, right.

Thanks for review, @sockeqwe and @ming13, you know how much I appreciate that 😽 

@vanniktech I'll keep it open for a day or two, it's not a functional update so there is no pressure in merging asap, you might have some comments I would assume :)

vanniktech commented:
Yeah wanted to wait until you address the feedback of others so we don't get too crazy here.

vanniktech commented on samples/redux/rxredux/src/main/java/com/lyft/domic/samples/redux/rxredux/signin/AndroidSignInView.kt:
```diff
@@ -0,0 +1,71 @@
+    override fun render(signInState: Observable<State>): Disposable {
+        val state = signInState
+                .replayingShare()
+                .observeOn(Schedulers.computation())
```
Doesn't this need to be the main thread? 4 lines below you do what looks like an operation that needs to run on the main thread.

artem-zinnatullin replied:
No, that's one of the major points of Domic, you're supposed to run your code off the main thread, Domic's rendering will take care of switching to main thread as efficient as we can

sockeqwe commented on samples/redux/rxredux/src/main/java/com/lyft/domic/samples/redux/rxredux/signin/AndroidSignInView.kt:
```diff
@@ -0,0 +1,71 @@
+
+    override fun render(signInState: Observable<State>): Disposable {
+        val state = signInState
+                .replayingShare()
```
Do you need `replayingShare()`?

artem-zinnatullin replied:
yeah, you need `share` or `replayingShare` to allow multiple streams down below in this method to observe same state

otherwise `state` Observable will run multiple times if it's Cold

sockeqwe commented on samples/redux/rxredux/src/main/java/com/lyft/domic/samples/redux/rxredux/signin/SignInStateMachine.kt:
```diff
@@ -0,0 +1,133 @@
+                        }
+            }
+
+
```
NIT: empty lines

sockeqwe commented on samples/redux/rxredux/src/main/java/com/lyft/domic/samples/redux/rxredux/signin/AndroidSignInView.kt:
```diff
@@ -0,0 +1,71 @@
+                    emailEditText
+                            .observe
+                            .textChanges
+                            .map { Action.ChangeEmail(it) },
```
```kotlin
emailEditText
             .observe
            .textChanges
            .map { Action.ChangeEmail(it) },
```
Duplicate?

artem-zinnatullin replied:
yis

sockeqwe commented on samples/redux/rxredux/src/main/java/com/lyft/domic/samples/redux/rxredux/signin/SignInStateMachine.kt:
```diff
@@ -0,0 +1,133 @@
+                    signInButtonEnabled = signInButtonEnabled
+            )
+        }
+        is Action.SignIn -> State.SigningIn(
```
You could just return previous `state` since `Action.SignIn` is handled in `signInSideEffect`.

sockeqwe commented on samples/redux/rxredux/src/main/java/com/lyft/domic/samples/redux/rxredux/signin/SignInStateMachine.kt:
```diff
@@ -0,0 +1,133 @@
+                password = state.password,
+                signInButtonEnabled = false
+        )
+        is Action.SigningIn -> State.SigningIn(
```
You can group `is Action.SignIn` and `is Action.SigningIn` together: 

```kotlin
when(action) {
 ...
  is Action.SignIn,
  is Action.SigningIn -> State.SigningIn( 
                 email = state.email,
                 password = state.password,
                 signInButtonEnabled = false
         )
  ...
}
```

artem-zinnatullin replied:
yeah, I was actually using multi `is` statement for `ChangeEmail` and `ChangePassword`, but quickly learned that Kotlin compiler doesn't narrow the type variance there and still requires me to check all the branches of `is` inside that block

gladly for this `SignIn` and `SigningIn` combination I don't need to look up into their fields

oldergod commented on samples/redux/rxredux/src/main/java/com/lyft/domic/samples/redux/rxredux/signin/SignInStateMachine.kt:
```diff
@@ -0,0 +1,126 @@
+
+
+    val state: Observable<State> = inputActions
+            .observeOn(computationScheduler)
```
Is that necessary?
With that the `signInButtonEnabled = false` will only apply to at least the next frame. The user could tap it many times?

artem-zinnatullin replied:
that let's to move computations that don't need run on main thread from main thread thus making it more capable for handling other stuff like animations

`observeOn(computationScheduler)` on its own doesn't mean that rendering will happen on next frame, frame window is normally 16.6ms, computation scheduler can finish this particular code within 1ms

However, rendering in Domic (with standard renderer implementation) always happens on next frame anyway, but as efficient as we can otherwise. 

Note that normally `Handler.post()` will give you similar behavior

This was discussed with Adam Powell (Android Framework UI lead and we agreed that one frame is nothing for non-game applications and is totally fine if  stable 60 or more fps is more important)

There is also api for manual rendering asap: #19

artem-zinnatullin replied:
yep, I know, you can always use API provided in #19, if you care about rendering in current frame (which most utility apps shouldn't really care about)

sockeqwe commented:
@artem-zinnatullin I think you should just merge this to master and people will / can provide further improvement by sending a PR ...

artem-zinnatullin commented:
Sounds good 👍
